renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
-                                                 ASMemory &result)
+uint64_t VulkanAccelerationStructureManager::GetSize_InitialState(ResourceId id,
+                                                                  const VkInitialContents &initial)
 {
-  const VkDeviceSize serialisedSize = SerialisedASSize(unwrappedAs);
-
-  const VkDevice d = m_pDriver->GetDev();
-  VkResult vkr = VK_SUCCESS;
-
-  // since this happens during capture, we don't want to start serialising extra buffer creates,
-  // leave this buffer as unwrapped
-  VkBuffer dstBuf = VK_NULL_HANDLE;
-
-  VkBufferCreateInfo bufInfo = {
-      VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
-      NULL,
-      0,
-      serialisedSize,
-      VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT |
-          VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT,
-  };
-
-  // we make the buffer concurrently accessible by all queue families to not invalidate the
-  // contents of the memory we're reading back from.
-  bufInfo.sharingMode = VK_SHARING_MODE_CONCURRENT;
-  bufInfo.queueFamilyIndexCount = (uint32_t)queueFamilyIndices.size();
-  bufInfo.pQueueFamilyIndices = queueFamilyIndices.data();
-
-  // spec requires that CONCURRENT must specify more than one queue family. If there is only one
-  // queue family, we can safely use exclusive.
-  if(bufInfo.queueFamilyIndexCount == 1)
-    bufInfo.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
-
-  vkr = ObjDisp(d)->CreateBuffer(Unwrap(d), &bufInfo, NULL, &dstBuf);
-  CHECK_VKR(m_pDriver, vkr);
+  const uint64_t infoSize = initial.accelerationStructureInfo->GetSerialisedSize();
+  const uint64_t serialisedASSize =
+      (sizeof(uint64_t) * 2) + initial.mem.size + WriteSerialiser::GetChunkAlignment();
 
-  m_pDriver->AddPendingObjectCleanup(
-      [d, dstBuf]() { ObjDisp(d)->DestroyBuffer(Unwrap(d), dstBuf, NULL); });
-
-  VkMemoryRequirements mrq = {};
-  ObjDisp(d)->GetBufferMemoryRequirements(Unwrap(d), dstBuf, &mrq);
-
-  mrq.alignment = RDCMAX(mrq.alignment, asBufferAlignment);
-
-  const MemoryAllocation readbackmem = m_pDriver->AllocateMemoryForResource(
-      true, mrq, MemoryScope::InitialContents, MemoryType::Readback);
-  if(readbackmem.mem == VK_NULL_HANDLE)
-    return false;
-
-  vkr = ObjDisp(d)->BindBufferMemory(Unwrap(d), dstBuf, Unwrap(readbackmem.mem), readbackmem.offs);
-  CHECK_VKR(m_pDriver, vkr);
-
-  const VkBufferDeviceAddressInfo addrInfo = {VK_STRUCTURE_TYPE_BUFFER_DEVICE_ADDRESS_INFO, NULL,
-                                              dstBuf};
-  const VkDeviceAddress dstBufAddr = ObjDisp(d)->GetBufferDeviceAddressKHR(Unwrap(d), &addrInfo);
+  return 128ULL + infoSize + serialisedASSize;
+}
 
-  VkCommandBuffer cmd = m_pDriver->GetInitStateCmd();
-  if(cmd == VK_NULL_HANDLE)
-  {
-    RDCERR("Couldn't acquire command buffer");
-    return false;
-  }
+bool VulkanAccelerationStructureManager::Serialise(WriteSerialiser &ser, ResourceId id,
```

Comment:
Please keep these serialise functions as templated over the serialiser type rather than explicitly specialising for read and write. I know in cases like this with initial contents there is not much commonality but every other serialise function is written like this, with conditional code for each mode, so I would not want to have an exception only for this case.